Resource aliases

dantleech opened the pull request:
This PR aims to allow registring aliases for resource _payloads_ and adds support for decorating the REST response.

For example:

``` javascript
            {
                "path": "\/foo",
                "repo_path": "\/foo",
                "children": [],
                "node": {
                    "jcr:primaryType": "nt:unstructured",
                    "jcr:mixinTypes": [
                        "phpcr:managed"
                    ],
                    "phpcr:class": "Symfony\\Cmf\\Bundle\\ResourceRestBundle\\Tests\\Resources\\TestBundle\\Document\\Article",
                    "phpcr:classparents": [],
                    "title": "Article 1",
                    "body": "This is my article"
                },
                "repository_alias": "phpcr_repo",
                "repository_type": "doctrine_phpcr",
                "payload_alias": "nt:unstructured",
                "_links": {
                    "self": {
                        "href": "\/api\/phpcr_repo\/foo"
                    }
                }
            }
                "phpcr:classparents": [],
                "title": "Article 2"
            },
        },
    },
}
```

``` yaml
cmf_resource_rest:
    payload_alias_map:
        article:
            repository: doctrine_phpcr
            type: "Symfony\\Cmf\\Bundle\\ResourceRestBundle\\Tests\\Resources\\TestBundle\\Document\\Article"
```

dantleech committed:
Resource aliases

dantleech commented on Serializer/EventSubscriber/ResourceSubscriber.php:
```diff
@@ -0,0 +1,60 @@
+
+        if ($object instanceof Resource) {
+            $visitor = $event->getVisitor();
+            $visitor->addData('repository', $this->registry->getName($object->getRepository()));
```
This serializer will add the extra data..

dantleech committed:
Made behat tests more stable

dantleech committed:
Support type

dantleech commented:
Updated, this should be ready now (pending review)

I have added the following:

```
                "repository_alias": "phpcr_repo",
                "repository_type": "doctrine_phpcr",
                "payload_alias": "nt:unstructured",
```

The `payload_alias` is either the node type (in the case of a PHPCR node) or a FQCN in the case of PHPCR-ODM. "Aliases" can be mapped as above to provide sane names. I guess the principle is the same as with XML namespaces.

dantleech commented:
/cc @WouterJ @dbu

wouterj commented on DependencyInjection/Configuration.php:
```diff
@@ -0,0 +1,43 @@
+        $treeBuilder = new TreeBuilder();
+        $treeBuilder->root('cmf_resource_rest')
+            ->children()
+                ->arrayNode('payload_alias_map')
```
requires `->fixXmlConfig`

dantleech replied:
Does it?

https://github.com/symfony-cmf/ResourceRestBundle/pull/8/files#diff-d89eb38c26b7d8b615be4d9f58b7dc5cR7

wouterj replied:
Yes, that XML is invalid. It is using an undefined `article` element. It has to be:

``` xml
<?xml version="1.0" charset="UTF-8" ?>
<container xmlns="http://symfony.com/schema/dic/services"
    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
    xmlns:cmf-resource-rest="http://cmf.symfony.com/schema/dic/cmf_resource_rest"
    xsi:schemaLocation="http://symfony.com/schema/dic/services http://symfony.com/schema/dic/services/services-1.0.xsd
        http://symfony.com/schema/dic/symfony http://symfony.com/schema/dic/symfony/symfony-1.0.xsd">
    <cmf-resource-rest:config>
        <cmf-resource-rest:payload-alias-map alias="article"
            type="Namespace\Article"
            repository="doctrine_phpcr_odm"
        />
    </cmf-resource-rest:config>
</container>
```

dantleech replied:
Hmm. Invalid why? If we change it then this would just be `alias` .. how would that look in YAML?

wouterj replied:
Yaml would be the same. The `->useAttributeAsKey` will transform the `alias` attribute in XML to the key of the array. In Symfony XML, prototyped arrays look like this:

``` xml
<alias id="article" type="..." />
<alias id="other_alias" type="..." />
```

``` yaml
alias_map:
    article: { type: '...' }
    other_alias: { type: '...' }
```

``` php
$root
    ->fixXmlConfig('alias', 'alias_map')
    ->children()
        ->arrayNode('alias_map')
            ->useAttributeAsKey('id')
            ->prototype('array')
                ->children()
                    ->scalarNode('type')->end()
                ->end()
            ->end()
        ->end() // alias_map
    ->end();
```

wouterj commented:
I can't find `Symfony\Cmf\Bundle\ResourceRestBundle\ResourceRest\PayloadAliasRegistry` anywhere in this PR (and travis can't either).

dantleech committed:
Added payloadaliasregistry

dantleech commented:
Added it. The travis builds are going to fail anyway as there are pending PRs in resource and resource-bundle.

dantleech commented on Serializer/EventSubscriber/ResourceSubscriber.php:
```diff
@@ -0,0 +1,67 @@
+use JMS\Serializer\EventDispatcher\ObjectEvent;
+use Puli\Repository\Api\Resource\Resource;
+use Symfony\Cmf\Component\Resource\RepositoryRegistryInterface;
+use Symfony\Cmf\Bundle\ResourceRestBundle\ResourceRest\PayloadAliasRegistry;
```
Remove unused statements

wouterj commented on DependencyInjection/Configuration.php:
```diff
@@ -0,0 +1,43 @@
+        return $treeBuilder;
+    }
+}
+
```
empty line

wouterj commented on ResourceRest/PayloadAliasRegistry.php:
```diff
@@ -0,0 +1,90 @@
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Cmf\Bundle\ResourceRestBundle\ResourceRest;
```
I'm -1 for such duplicated namespaces, we should move this to the root of the bundle instead in that case imo

dantleech replied:
Well, I do this with the aim of creating a component .. but in this case we could move it to the bundle root.

wouterj replied:
Would it be usefull to extract a component out of this bundle?

dantleech replied:
I don't think so.

wouterj commented on ResourceRest/PayloadAliasRegistry.php:
```diff
@@ -0,0 +1,90 @@
+        RepositoryRegistryInterface $repositoryRegistry,
+        array $aliases = array()
+    )
+    {
```
this has to be `) {`

wouterj commented on ResourceRest/PayloadAliasRegistry.php:
```diff
@@ -0,0 +1,90 @@
+
+        if (!isset($this->aliasesByRepository[$repositoryType])) {
+            throw new \RuntimeException(sprintf(
+                'No repositories registered with alias "%s". I know about "%s"',
```
"I know about" -> "Known aliases:"

dantleech committed:
Fixed XML config

dantleech committed:
Fixed tests

dantleech committed:
CS fixes

dantleech committed:
Decorator support

dantleech commented on Decorator/DecoratorInterface.php:
```diff
@@ -0,0 +1,35 @@
+     *     $context->addData('foobar', 'Some value');
+     *
+     * @param Context Serialization context
+     * @param resource The resource being serialized
```
Resource

dantleech commented on DependencyInjection/Configuration.php:
```diff
@@ -0,0 +1,49 @@
+                    ->end()
+                ->end()
+                ->arrayNode('decorator_map')
+                    ->useAttributeAsKey('repository')
```
fix XML config

dantleech commented on Tests/Unit/DependencyInjection/fixtures/config.xml:
```diff
@@ -0,0 +1,9 @@
+        http://symfony.com/schema/dic/symfony http://symfony.com/schema/dic/symfony/symfony-1.0.xsd">
+    <cmf_resource_rest:config>
+        <cmf_resource_rest:payload_alias name="article" type="Namespace\Article" repository="doctrine_phpcr_odm" />
+    </cmf_resource_rest:config>
```
Add decorator map test

dantleech commented:
Updated to include support for decorating the response. In doing so I removed the `payload` element from the response and made it into a decorator.

Next step I think will be making a SonataAdmin decorator.

dantleech commented on Decorator/PayloadDecorator.php:
```diff
@@ -0,0 +1,32 @@
+    public function decorate(Context $context, Resource $resource)
+    {
+        $visitor = $context->getVisitor();
+        $visitor->addData('payload', $context->accept($resource->getPayload()));
```
@WouterJ @dbu is this even decoration? or should we use a different term?

dantleech replied:
I guess the Enhancer might be better in fitting with the RoutingBundle.

dantleech committed:
Renamed decorator to enhancer

dantleech commented:
Test failing due to deprecation errors, see: https://github.com/symfony-cmf/Testing/pull/93
